cpp/onedal/array.hpp

```diff
@@ -18,79 +18,170 @@
+        return std::holds_alternative<T*>(data_) && (get_mutable_data() != nullptr);
+    }
 
-    template <typename Deleter>
-    explicit array(T* data, std::int64_t size, Deleter&& deleter)
-        : data_(data, deleter),
-          size_(size) {}
+    array& unique() {
+        if (is_data_owner() || size_ == 0) {
+            return *this;
+        } else {
+            auto immutable_data = get_data();
+            auto copy_data = new T[size_];
 
-    template <typename U>
-    array(const array<U>& other)
-        : data_(other.data_),
-          size_(other.size_) {}
+            for (std::int64_t i = 0; i < size_; i++) {
+                copy_data[i] = immutable_data[i];
+            }
 
-    T* get_pointer() const noexcept {
-        return data_.get();
+            reset(copy_data, size_, default_delete{});
+            return *this;
+        }
     }
 
-    std::int64_t get_size() const noexcept {
+    std::int64_t get_size() const {
         return size_;
     }
 
+    std::int64_t get_capacity() const {
+        return capacity_;
+    }
+
+    bool is_data_owner() const {
+        if (data_owned_ptr_ == nullptr) {
+            return false;
+        } else if (auto ptr_val = std::get_if<T*>(&data_)) {
+            return *ptr_val == data_owned_ptr_.get();
+        } else if (auto ptr_val = std::get_if<const T*>(&data_)) {
+            return *ptr_val == data_owned_ptr_.get();
+        } else {
+            return false;
+        }
+    }
+
     void reset() {
-        data_.reset();
+        data_owned_ptr_.reset();
+        data_ = std::variant<T*, const T*>();
         size_ = 0;
+        capacity_ = 0;
+    }
+
+    void reset(std::int64_t size) {
+        data_owned_ptr_.reset(new T[size], default_delete{});
+        data_ = data_owned_ptr_.get();
+        size_ = size;
+        capacity_ = size;
+    }
+
+    template <typename Deleter>
+    void reset(T* data, std::int64_t size, Deleter&& deleter) {
+        // TODO: check input parameters
+        data_owned_ptr_.reset(data, std::forward<Deleter>(deleter));
+        data_ = data_owned_ptr_.get();
+        size_ = size;
+        capacity_ = size;
     }
 
-    const T& operator [](std::int64_t index) const noexcept {
-        return get_pointer()[index];
+    template <typename U = T*>
+    void reset_not_owning(U data = nullptr, std::int64_t size = 0) {
+        data_ = data;
+        size_ = size;
     }
 
-    T& operator [](std::int64_t index) noexcept {
-        return get_pointer()[index];
+    void resize(std::int64_t size) {
+        if (is_data_owner() == false) {
+            throw std::runtime_error("cannot resize array with non-owning data");
+        } else if (size <= 0) {
+            reset_not_owning();
+        } else if (get_capacity() < size) {
+            T* new_data = new T[size];
+            std::int64_t min_size = std::min(size, get_size());
+
+            for (std::int64_t i = 0; i < min_size; i++) {
+                new_data[i] = (*this)[i];
```

Comment:
As we do not know the actual type of `T`, at this point copy constructor (what if `T` is `std::vector`) could throw an exception, which results in memory leak in `new_data` .